pkg/kubelet/kubelet_pods.go

```diff
@@ -1635,7 +1635,15 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
 				}
 				// Secondary IP is not set #105320
 				if len(hostIPs) == 2 && len(s.PodIPs) == 1 {
-					s.PodIPs = append(s.PodIPs, v1.PodIP{IP: hostIPs[1].String()})
+					currIsIPv6 := utilnet.ParseIPSloppy(s.PodIPs[0].IP).To4() == nil
```

Comment:
I see, if the first hostIP returns IPA, and the subsequent request returms IPB,IPA in different order, it will add { IPA, IPA}

However, there are a lot of places in the code that depend on the order of the returned address, that is why we have
```
func (kl *Kubelet) sortPodIPs(podIPs []string) []string {
```

that is called from 
```
convertStatusToAPIStatus
```

/cc @danwinship he has a keep to try to organize this

Reply:
I also don't understand why the Node.Status.Addresses (which is the underlying slice data structure that `hostIPs` represents) would inconsistently represent a well-ordered set.

Reply:
I don't remember if we had something about the order, typically this may be added by an external controller, but in this case it seems that the external controller is adding the addresses in reverse order

Reply:
If the problem is the order, then doing this additive mutation should really be transformed to just look for net new IPs, rather than assuming that the particular pathology is restricted to swapping the order of ipv4 and ipv6 IPs _in particular_. Does that make sense?

I think a solution like this is more general, and preferred:

```
$ git diff pkg/kubelet/kubelet_pods.go
diff --git a/pkg/kubelet/kubelet_pods.go b/pkg/kubelet/kubelet_pods.go
index fbc9104c428..85434f823b8 100644
--- a/pkg/kubelet/kubelet_pods.go
+++ b/pkg/kubelet/kubelet_pods.go
@@ -1635,7 +1635,11 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
                                }
                                // Secondary IP is not set #105320
                                if len(hostIPs) == 2 && len(s.PodIPs) == 1 {
-                                       s.PodIPs = append(s.PodIPs, v1.PodIP{IP: hostIPs[1].String()})
+                                       for _, ip := range hostIPs {
+                                               if ip.String() != s.PodIP {
+                                                       s.PodIPs = append(s.PodIPs, v1.PodIP{IP: ip.String()})
+                                               }
+                                       }
                                }
                        }
                }
```

This passes your new UT @lzhecheng 

cc @CecileRobertMichon who is also tracking this IPV6 + OOT cloud-provider-azure pathology

Reply:
@aojea hello, do you suggest using `sortPodIPs()` for `hostIPs`? I think it may also work.
Yes, IPs are added by external controller, but `kl.getHostIPsAnyWay()` doesn't ensure the order right?

Reply:
@jackfrancis I also thought about your suggested solution. But I think the key is: In PodIPs, IPs of same IP family should be only one. What do you think?

Reply:
If indeed there are strict combinatoric requirements (only ipv4 or ipv6 or ipv4+ipv6 but not ipv4+ipv4 or ipv6+ipv6 or > 2 IPs) then IMO that validation should be discrete and separate from this “add a secondary IP to an existing IP” flow. Dealing with those combinatoric requirements *and* dealing with reordering/merging in this single flow makes the code almost unreadable, it seems to me.

My 2¢

Reply:
There's indeed such requirement, which is why the OOT + IPv6 issue comes.
I think reordering/merging is fine while picking a value from a fixed index seems dangerous. It's my opinion and let's have more people taking a look.

Reply:
sgtm

Reply:
I would also vote for that approach. Is someone working on that?

Reply:
@lzhecheng ? 😄

Reply:
I'm happy to solve this issue with the proper approach. But please correct me if I'm wrong.
The order here involves the hostIPs set last time, which can be got with `s.PodIPs`. So `getHostIPsAnyWay()` needs something like last time IP family as an input. Even if we do it this way, we cannot simply add the second IP into `PodIPs` because that will leads to the problem I met now (`<IPv6>, <getHostIPsAnyWay>`), so the rest of code still needs to check IP family.
There's another approach which avoids rest of code dealing with these things, which is overwriting PodIPs with host IPs. Can we do it? I'm not sure since I saw your comment mentioning "immutable once set". I don't know if overwriting will lead to other problems.

Reply:
let me try to find some time this week to get to this, this part is very brittle and delicate and I'd like to be completely sure before I suggest anything

Reply:
> I still feel like #106715 was wrong, and the right fix is "wait for your network to settle before you start kubelet".

to be more specific, it should be "wait for your cloud-controller-manager to write the node.status.Addresses and in the right order?"

Reply:
@danwinship we have some expectations on the order and nodeIP is the source of truth

https://github.com/kubernetes/kubernetes/pull/97979 

> kubelet is assuming that the IPs are returned in preference order

In this case the problem is the controller is returning an order that doesn't match what we defined as the primary IP family of the kubelet.

Reply:
Ok, the order of Node IPs is cloud-provider's responsibility and I will follow up this issue in cloud-provider, Node side.
I think your suggestion is good and will update the PR

Reply:
@aojea Hello! Do you expect other changes from my side?

Reply:
sorry, went on vacation, Dan made a very detailed comment about some other edge cases that we may document and warn